packages/worker/src/worker-options.ts

```diff
@@ -411,13 +416,100 @@ export function addDefaultWorkerOptions(options: WorkerOptions): WorkerOptionsWi
+      // Maximum size (in bytes) of memory usage allowed by this container. 'max' if no constraint.
+      // Examples:
+      //  - 536870912 => 512 Mb
+      //  - max => no constraint
+      const memoryMax = Number(tryReadFileSync('/sys/fs/cgroup/memory.max'));
+      if (!isNaN(memoryMax)) resources.memory = Math.min(memoryMax, resources.memory);
+
+      // Maximum size (in bytes) of swap usage allowed by this container. The reported size EXCLUDES the size of memory itself.
+      // For example, assuming `docker run --memory 512mb --memory-swap 768mb`, memory.swap.max will report 256mb
+      // Examples:
+      //  - 268435456 => 256 Mb
+      //  - max => swap allowed but unconstrained
+      const swapMax = tryReadFileSync('/sys/fs/cgroup/memory.swap.max');
+      if (swapMax) {
+        if (swapMax === 'max')
+          resources.swap = resources.memory; // Swap is allowed, but no size was specified. Default to memory size
+        else if (!isNaN(Number(swapMax))) resources.swap = Number(swapMax);
+      }
+    }
+
+    // Check for v1 style cgroups: /sys/fs/cgroup/memory/memory.limit_in_bytes
+    if (fs.existsSync('/sys/fs/cgroup/memory/memory.limit_in_bytes')) {
+      // Maximum size (in bytes) of memory usage allowed by this container. '9223372036854771712' if no constraint
+      // Examples:
+      //  - 536870912 => 512 Mb
+      //  - 9223372036854771712 => swap allowed but unconstrained
+
+      const memoryMax = Number(tryReadFileSync('/sys/fs/cgroup/memory/memory.limit_in_bytes'));
+      if (!isNaN(memoryMax)) resources.memory = Math.min(memoryMax, resources.memory);
+
+      // Maximum size (in bytes) of swap usage allowed by this container. The reported size INCLUDES the size of memory itself.
+      // For example, assuming `docker run --memory 512mb --memory-swap 768mb`, memory.memsw.limit_in_bytes will report 768mb.
+      // File might also not exists if system does not support swap.
+      // Defaults to '9223372036854771712' if no constraint.
+      // Examples:
+      //  - 536870912 => 512 Mb
+      //  - 9223372036854771712 => no constraint
+      const swapMax = Number(tryReadFileSync('/sys/fs/cgroup/memory/memory.memsw.limit_in_bytes'));
+      if (!isNaN(swapMax)) {
+        if (swapMax === 9223372036854771712)
+          resources.swap = resources.memory; // Swap is allowed, but no size was specified. Default to memory size
+        else resources.swap = swapMax - resources.memory; // memory.memsw.limit_in_bytes includes mem+swap
+      }
+    }
+  }
+
+  if (resources.heap > resources.memory * 0.5) {
+    Runtime.instance().logger.warn(
+      `node's heap size limit is over 50% of available memory ` +
+        `(heap size limit = ${resources.heap}, system memory = ${resources.memory}) ` +
+        `This might degrade performances and could result in OOM errors. ` +
+        `For optimal performance, try setting '--max-old-space-size' between 25% and 50% of system memory.`
```

Comment:
I agree with what you said about having a conservative default, but I wonder if we can give a simple single % rec for them to start at, and if most of workflow cache is in heap, then recommending higher than 50%.